### PR TITLE
QB-1574: update to TM calls, refactor ws

### PR DESCRIPTION
**Pull request by BoThe1K**
Tests: - https://github.com/stratosnet/stratos-evm-tests

### CODE REVIEW

File: rpc/backend/utils.go [hunk 1]
```diff
@@ -170,64 +173,34 @@ func (b *Backend) SetTxDefaults(args evmtypes.TransactionArgs) (evmtypes.Transac
...
-				continue
-			}
-			if sender == accAddr {
-				nonce++
-			}
-		}
-	}
-
-	return nonce, nil
+	return acc.GetSequence() + pendingNonce
```
Comment (Xiong-stratos) [inline rpc/backend/utils.go:191]: Why return acc.GetSequence() + pendingNonce?

File: rpc/types/utils.go [hunk 2]
```diff
@@ -40,37 +49,116 @@ func RawTxToEthTx(clientCtx client.Context, txBz tmtypes.Tx) ([]*evmtypes.MsgEth
...
+		UncleHash:   common.Hash{},
+		Coinbase:    common.BytesToAddress(header.ProposerAddress.Bytes()),
 		Root:        common.BytesToHash(header.AppHash),
-		TxHash:      txHash,
-		ReceiptHash: ethtypes.EmptyRootHash,
+		TxHash:      common.BytesToHash(header.DataHash),
+		ReceiptHash: common.Hash{},
 		Bloom:       bloom,
-		Difficulty:  big.NewInt(0),
+		Difficulty:  big.NewInt(1), // NOTE: Maybe move to some constant?
```
[the remark this answers is not in the document]
Comment (BoThe1K) [reply rpc/types/utils.go:98]: @alexstratos so agreed on 1 or changing to `len(commit.Signatures)`?

[the remark this answers is not in the document]
Comment (alexstratos) [reply rpc/types/utils.go:116]: agreed on 1

**Review by alexstratos — APPROVED** [review]

**Comment by BoThe1K** [comment]
PR where state could be broken (for a new genesis dep):
https://github.com/stratosnet/stratos-chain/pull/218

File: x/evm/keeper/state_transition.go [hunk 3]
```diff
@@ -122,6 +124,9 @@ func (k Keeper) VMConfig(ctx sdk.Context, msg core.Message, cfg *types.EVMConfig
 //  2. The requested height is from an previous height from the same chain epoch
 //  3. The requested height is from a height greater than the latest one
 func (k Keeper) GetHashFn(ctx sdk.Context) vm.GetHashFunc {
+	cache := make(map[int64]common.Hash)
```
Comment (alexstratos) [inline x/evm/keeper/state_transition.go:127]: seems we never release the cache in any cases, maybe worth clean up after certain size?

Comment (BoThe1K) [reply x/evm/keeper/state_transition.go:127]: I just compared with https://github.com/ethereum/go-ethereum/blob/7d29fff415c9ad08ebb34e5052a89585ce60eee2/core/evm.go#L81
they also not clearing, but not sure that we should clear as leak here only on total count of blocks + N upcomming and only if smart contract has block.blockhash call, so should not be a problem. I could try to change from the map to a slice, should be less size.

**Pull request merged**